airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/SecretsRepositoryWriter.java

```diff
@@ -0,0 +1,268 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class takes secrets as arguments but never returns a secrets as return values (even the ones
+ * that are passed in as arguments). It is responsible for writing connector secrets to the correct
+ * secrets store and then making sure the remainder of the configuration is written to the Config
+ * Database.
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+public class SecretsRepositoryWriter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SecretsRepositoryWriter.class);
+
+  private static final UUID NO_WORKSPACE = UUID.fromString("00000000-0000-0000-0000-000000000000");
+
+  private final ConfigRepository configRepository;
+  private final JsonSchemaValidator validator;
+  private final Optional<SecretPersistence> longLivedSecretPersistence;
+  private final Optional<SecretPersistence> ephemeralSecretPersistence;
+
+  public SecretsRepositoryWriter(final ConfigRepository configRepository,
+                                 final Optional<SecretPersistence> longLivedSecretPersistence,
+                                 final Optional<SecretPersistence> ephemeralSecretPersistence) {
+    this(configRepository, new JsonSchemaValidator(), longLivedSecretPersistence, ephemeralSecretPersistence);
+  }
+
+  @VisibleForTesting
+  SecretsRepositoryWriter(final ConfigRepository configRepository,
+                          final JsonSchemaValidator validator,
+                          final Optional<SecretPersistence> longLivedSecretPersistence,
+                          final Optional<SecretPersistence> ephemeralSecretPersistence) {
+    this.configRepository = configRepository;
+    this.validator = validator;
+    this.longLivedSecretPersistence = longLivedSecretPersistence;
+    this.ephemeralSecretPersistence = ephemeralSecretPersistence;
+  }
+
+  private Optional<SourceConnection> getOptionalSourceConnection(final UUID sourceId) throws JsonValidationException, IOException {
+    try {
+      return Optional.of(configRepository.getSourceConnection(sourceId));
+    } catch (final ConfigNotFoundException e) {
+      return Optional.empty();
+    }
+  }
+
+  // validates too!
+  public void writeSourceConnection(final SourceConnection source, final ConnectorSpecification connectorSpecification)
+      throws JsonValidationException, IOException {
+
+    final var previousSourceConnection = getOptionalSourceConnection(source.getSourceId())
+        .map(SourceConnection::getConfiguration);
+
+    // strip secrets
+    final JsonNode partialConfig = statefulUpdateSecrets(
+        source.getWorkspaceId(),
+        previousSourceConnection,
+        source.getConfiguration(),
+        connectorSpecification);
+    final SourceConnection partialSource = Jsons.clone(source).withConfiguration(partialConfig);
+
+    // validate partial to avoid secret leak issues.
```

Comment:
So this would make the cloud version not include the coordinate, but only `*****`? Seems fine for now, but we will likely want to change that in the future when we support user secret stores on cloud.

Reply:
i don't think i follow. this is just doing a validation.